sycl/doc/extensions/proposed/sycl_ext_oneapi_global_defaults.asciidoc

```diff
@@ -0,0 +1,982 @@
+----
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename KernelName, int Dimensions,
+          typename Properties, typename... Args>
+void parallel_for(launch_config<sycl::range<Dimensions>, Properties> c,
+                  const sycl::kernel& k, Args&& args...);
+
+}
+----
+!====
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+parallel_for(get_current_device().ext_oneapi_get_default_queue(), c, k, args)
+----
+
+|====
+
+
+==== ND-range kernels
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename KernelName, int Dimensions,
+          typename KernelType, typename... Reductions>
+void nd_launch(sycl::nd_range<Dimensions> r,
+               const KernelType& k, Reductions&&... reductions);
+
+}
+----
+!====
+_Constraints_: The parameter pack consists of 0 or more objects created by the
+`sycl::reduction` function.
+
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+nd_launch(get_current_device().ext_oneapi_get_default_queue(), r, k, reductions)
+----
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename KernelName, int Dimensions,
+          typename Properties,
+          typename KernelType, typename... Reductions>
+void nd_launch(launch_config<sycl::nd_range<Dimensions>, Properties> c,
+               const KernelType& k, Reductions&&... reductions);
+
+}
+----
+!====
+_Constraints_: The parameter pack consists of 0 or more objects created by the
+`sycl::reduction` function.
+
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+nd_launch(get_current_device().ext_oneapi_get_default_queue(), c, k, reductions)
+----
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename KernelName, int Dimensions, typename... Args>
+void nd_launch(sycl::nd_range<Dimensions> r,
+               const sycl::kernel& k, Args&&... args);
+
+}
+----
+!====
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+nd_launch(get_current_device().ext_oneapi_get_default_queue(), r, k, args)
+----
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename KernelName, int Dimensions,
+          typename Properties, typename... Args>
+void nd_launch(launch_config<sycl::nd_range<Dimensions>, Properties> c,
+               const sycl::kernel& k, Args&& args...);
+
+}
+----
+!====
+_Effects_: The same result as computed by:
+
+[source, c++]
+----
+nd_launch(get_current_device().ext_oneapi_get_default_queue(), c, k, args)
+----
+
+|====
+
+
+==== Memory operations
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+void memcpy(void* dest, const void* src, size_t numBytes);

```

Comment:
I predict that many users will be confused, thinking this is a synchronous operation.  For example, `syclex::malloc_shared` and `syclex::memcpy` are both memory operations.  How are users supposed to know that the first is synchronous while the second is asynchronous?

Maybe we should add the suffix "_async" to all the functions that are asynchronous?

Reply:
I'm not sure how I feel about the idea of an "_async" suffix at this point.  I'm a little worried that explicitly calling out that some functions are asynchronous would introduce the opposite problem to what you're trying to solve: because functions like `submit` and `parallel_for` aren't called `submit_async` and `parallel_for_async`, a new developer might conclude they're synchronous.

(Aside: I would be okay with introducing an "_async" suffix if that's the direction C++ takes for its asynchronous STL functions, but then we'd need to roll out the suffix everywhere and deprecate the previous function names.)

Another potential fix would be to say that all these default queue shorthands are blocking...?  I guess then there's a confusion that `sycl::memcpy(dest, src, numBytes)` is synchronous but `sycl::memcpy(q, dest, src, numBytes)` is asynchronous...

Reply:
> I'm a little worried that explicitly calling out that some functions are asynchronous would introduce the opposite problem to what you're trying to solve: because functions like `submit` and `parallel_for` aren't called `submit_async` and `parallel_for_async`, a new developer might conclude they're synchronous.

I was thinking that we could add "_async" to all of the asynchronous functions in the [sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_enqueue_functions.asciidoc) extension, but we would leave the functions in SYCL 2020 named as they are now.